### PR TITLE
🚨 [lightglue] fix: matches order changed because of early stopped indices

### LINKED ISSUE #171: Batch Inference Code on Transformers
Hi! When running batch inference code, the order of the returned matching_scores is incorrect. For example, if the input batch order is (0, 1, 2), the returned result might be (0, 2, 1). This is due do the early stopping mechanism of Transformers Layers in the paper, which shuffles the batch index. Yet the code of recovering the batch index is incorrect.

Here is the Bug-Fixed code.
at modeling_lightglue.py, lin639
`idx_map = {early_stops_indices[i].item(): i for i in range(early_stops_indices.size()[0])} # debug`
`idx_new = [idx_map[i] for i in range(early_stops_indices.size()[0])] # debug`
`matches, matching_scores, final_pruned_keypoints_indices, final_pruned_keypoints_iterations = ( `
   ` tensor[idx_new] # tensor[early_stops_indices], debug`
    `for tensor in [`
    `matches,`
    `matching_scores,`
    `final_pruned_keypoints_indices,`
    `final_pruned_keypoints_iterations,]`
 ` )`

### PR DESCRIPTION
**Pull request by sbucaille**
# What does this PR do?

Fixes https://github.com/cvg/LightGlue/issues/171#issuecomment-3284295107
A bug is present in LightGlue when using batching.
The way early stopped indices were handled made order of matches change.
Example :
```python
# Input being
[[image2, image0], [image2, image0], [image1, image1]]
# Considering image2 and image0 very dissimilar and image1 is perfect match to itself
# Output would be similar to
[50, 900, 50]
# instead of
[50, 50, 900]
```
This PR fixes the bug by rearranging the early stopped indices which are used to rearrange the matches and other tensors

## Before submitting
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you write any new necessary tests?

## Who can review?

@qubvel

### CODE REVIEW

File: src/transformers/models/lightglue/modeling_lightglue.py [hunk 1]
```diff
@@ -628,6 +628,8 @@ def _concat_early_stopped_outputs(
         matching_scores,
     ):
         early_stops_indices = torch.stack(early_stops_indices)
+        # Rearrange tensors to have the same order as the input batch
+        early_stops_indices = early_stops_indices[early_stops_indices[torch.arange(early_stops_indices.shape[0])]]
```
Comment (qubvel) [inline src/transformers/models/lightglue/modeling_lightglue.py]: I would split this into two lines of code for easier following and debug, smth like

```suggestion
        # Rearrange tensors to have the same order as the input batch
        ids = torch.arange(early_stops_indices.shape[0])
        order_indices = early_stops_indices[ids]
        early_stops_indices = early_stops_indices[order_indices]
```

**Review by qubvel — COMMENTED** [review]
Thanks for the fix and adding a test! Very nice

**Comment by qubvel** [comment]
run-slow: lightglue

**Comment by qubvel** [comment]
@sbucaille it seems that the integration test failed, can you please have a look?

**Comment by sbucaille** [comment]
Oh my... the bug was in front of my eyes since the beginning, what a shame 😐 should be fixed !

**Comment by sbucaille** [comment]
@qubvel gentle bump

**Comment by qubvel** [comment]
run-slow: lightglue

**Review by qubvel — APPROVED** [review]
